interfaces/builtin/dbus_bind.go

```diff
@@ -0,0 +1,380 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+var dbusBindPermanentSlotAppArmorShared = []byte(`
+# Description: Allow owning a name on DBus public bus
+
+#include <abstractions/###DBUS_BIND_ABSTRACTION###>
+
+# register on DBus
+dbus (send)
+    bus=system
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+dbus (send)
+    bus=system
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="GetConnectionUnix{ProcessID,User}"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+`)
+
+var dbusBindPermanentSlotAppArmorIndividual = []byte(`
+# bind to a well-known DBus name: ###DBUS_BIND_NAME###
+dbus (bind)
+    bus=###DBUS_BIND_BUS###
+    name=###DBUS_BIND_NAME###,
+
+# Allow us to talk to dbus-daemon
+dbus (receive)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+dbus (send)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    interface=org.freedesktop.DBus.Properties
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+`)
+
+var dbusBindPermanentSlotAppArmorIndividualClassic = []byte(`
+# allow unconfined clients talk to ###DBUS_BIND_NAME### on classic
+dbus (receive)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    peer=(label=unconfined),
+`)
+
+var dbusBindPermanentSlotSecComp = []byte(`
+# Description: Allow owning a name on DBus public bus
+
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+var dbusBindConnectedSlotAppArmorIndividual = []byte(`
+# Description: Allow DBus consumer to connect to ###DBUS_BIND_NAME###
+
+# Communicate with the well-known named DBus service
```

Comment:
First of, this PR represents an exploration of what we could do, why that might be good and the considerations around it.

This is implementing a generic dbus interface. The bind allows binding to a given location, as you mentioned, and we could stop there to unblock gnome apps which I think is what people were expecting. However, I implemented that and got to thinking "why are these gnome apps doing this?" and came to the conclusion that it was so that they could integrate with the shell and the user's session, so I added 'unconfined' rules for that. Then it occurred to me that if we stopped there we would have a bug request next week or next month saying "gnome apps can't interoperate with each other" so I implemented how to connect snaps together in the manner described.

Now, interfaces are a form of contract and this contract is extremely loose-- all it says is once connected you can communicate with whatever methods they want. In practice this is 'probably' ok because this will be used with freedesktop.org standards and stuff 'probably' won't break. But 'probably' is again not a strong contract. I mentioned that we could explore the concept of auto-connecting based on if the publisher matches, and I think if we use this generic approach we should do this (much like with the content interface).

So as an exploration, this is interesting and while I was implementing this I was inspired by the content interface since I think there are many parallels. Where the content interface says "here is a location with a bunch of stuff I assume you know what to do with", the dbus-bind interface says "here is a well-known dbus name with an API I assume you know how to use". If we treat dbus-bind similarly in terms of auto-connecting, I was able to imagine a gnome ecosystem really starting to take off-- ie, gnome developers have a content snap and they develop gnome snaps that autoconnect to this content snap, and those developers may use dbus-bind in their snaps and they autoconnect. Beyond gnome, I thought this might be useful for KDE (snaps from KDE devs), Ubuntu Personal (from Canonical devs), etc. Also, like with the content sharing interface, the user is allowed to manually connect snaps from different publishers at their discretion.

I'm happy to adjust this PR as necessary to scale it back, put more constraints, go another direction, etc as we continue to explore this concept.